registry/handlers/images.go

```diff
@@ -83,6 +91,64 @@ func (imh *imageManifestHandler) GetImageManifest(w http.ResponseWriter, r *http
 		return
 	}
 
+	supportsSchema2 := false
+	supportsManifestList := false
+	if acceptHeaders, ok := r.Header["Accept"]; ok {
+		for _, mediaType := range acceptHeaders {
+			if mediaType == schema2.MediaTypeManifest {
+				supportsSchema2 = true
+			}
+			if mediaType == manifestlist.MediaTypeManifestList {
+				supportsManifestList = true
+			}
+		}
+	}
+
+	// Only rewrite schema2 manifests when they are being fetched by tag.
+	// If they are being fetched by digest, we can't return something not
+	// matching the digest.
+	if schema2Manifest, isSchema2 := manifest.(*schema2.DeserializedManifest); imh.Tag != "" && isSchema2 && !supportsSchema2 {
```

Comment:
Hmm, this approach would be possible, but it would involve passing lots of extra information to `ManifestService.Get` (with functional args?) so it can decide whether a conversion is necessary and possibly perform the conversion:

- tag
- whether the client supports schema2
- whether the client supports manifest lists
- trust key

So I'm up for giving it a try if you want, but my immediate reaction is that needing to pass all this infromation down through `ManifestService.Get` makes it cleaner to do the conversion here.